### ISSUE TITLE
Evaluate one image performance 7 times slower than using opencv dnn moudle, why ?

### ISSUE THREAD

**Issue by died** [Microsoft/CNTK#2556]
I tried using CNTK framework to train/evaluate model, it was fast. But when I want to evaluate only one image and get the result, it seems too slow for production using.

With pre-trained model (ResNet50_ImageNet_CNTK.model) in C#/python using GPU as device, it take more than 1100ms to evaluate one image, but I tried evaluate same image with Caffee model in OpenCv Dnn Python, it only take about 150ms.

I don't know why CNTK evaluate with bulk data can be fast but one image seems too slow,  or it need some trick to make it faster ?

I put my test code on https://github.com/died/CntkEvaluatePerformance , more detail in here too.

Hope someone can help for it, thank you.

**Comment by cha-zhang** [comment]
There is typically overhead when evaluating the first image due to model validation. Could you try multiple evaluations and see if it gets faster?

**Comment by died** [comment]
done few more test, create cntk MinibatchSource and read image one by one to evaluate, first two evaluate round was slow (about 1sec and 14sec), but follow up evaluate was very fast, increase batch size got almost same result, first two test still way too slow.
![img](https://i.imgur.com/v8sipIY.png)

Any tip can reduce the overhead time, like warm up or something else  ?


________________________________
part of test code.

            Function model = Function.Load(modelFile, device);
            var outputVar = model.Outputs.Single(variable => string.Equals(variable.Name, outputName));

            MinibatchSource minibatchSource = CreateMinibatchSource("imglist.txt", new [] { 224, 224, 3 }, 1000);
            var featureStreamInfo = minibatchSource.StreamInfo("image");

            int imgbatch = 1, totalCount = 0;
            while (true)
            {
                var minibatchData = minibatchSource.GetNextMinibatch((uint)imgbatch, device);

                if (minibatchData == null) break;
                totalCount += imgbatch;
                if (totalCount > 6) break;

                //setting
                var inputDataMap = new Dictionary<Variable, Value> { { model.Arguments[0], minibatchData[featureStreamInfo].data } };
                var outputDataMap = new Dictionary<Variable, Value> { { outputVar, null } };

                Stopwatch sw = new Stopwatch();
                sw.Start();

                //evaluate result            
                model.Evaluate(inputDataMap, outputDataMap, device);

                sw.Stop();
                Console.WriteLine($"Evaluate cost {sw.ElapsedMilliseconds} ms");

                //get result
                var outputVal = outputDataMap[outputVar];
                var actual = outputVal.GetDenseData<float>(outputVar);

                //get top 3 label
                foreach (var ans in actual)
                {
                    var actualLabels = ans.Select((x, index) => new { label = index, value = x }).ToList();
                    var list = actualLabels.OrderByDescending(x => x.value).Take(3);
                    foreach (var v in list)
                    {
                        Console.WriteLine($"Label:[{v.label}]{LabelName[v.label]}\tValue:{v.value}");
                    }
                }
            }


BTW, minibatchSource.GetNextMinibatch can't get null in right time when all batch run over, need use totalcount to exit while or the loop will keep running for a few time , dunno it's bug or not.

**Comment by cha-zhang** [comment]
In CNTK we have autotuning for determining which convolution to be used during execution. This happens in the second minibatch. 

Autotuning consumes extra memory, thus in the first minibatch we use zero/low memory convolution methods to make sure data buffers are allocated first. Then during the second minibatch we perform autotuning to find the fastest algorithm. Starting the third minibatch things will run very fast. 

You may consider feeding some dummy data with same dimension as your input, and treat first two minibatches as model initialization.

**Comment by died** [comment]
@cha-zhang ok I got it, thank you for explain how and why CNTK running like that, now I have a clear way to try use it on production, it help me a lot!!
